### DISCUSSION TITLE
Allow an affected target to run only on directly modified projects without walking the dependency graph

### DISCUSSION THREAD

**Discussion by meDavidNS** [nrwl/nx#22566] [Feature Requests]
- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md))

## Description
I would like to execute a target on all directly affected projects in the nx workspace. My usecase is specifically for linting, but could be usefull for formatting or docs generation as well.

## Motivation
For linting I would like to run the target only on directly affected projects, unlike unit tests or e2e tests, reducing the amount of time required to run `nx affected -t lint`. For linting it doesn't make sense to walk the dependency trying to executing the same target on other projects that depend on the directly modified project. 

## Suggested Implementation
For the `nx affected` command implement a --direct or --depth option.

Please let me know if you would welcome a PR.

**Comment by FrozenPandaz** [comment]
For lint in particular, it's generally not the slowest part of the pipeline so it's less important. If the lint tasks do not have any inputs which start with `^` though they will run, they would be cached. This is how the `lint` tasks are configured by default.

If we were to add something we would need to design it out further.

**Reply by meDavidNS** [reply]
My case might be a bit different. I use nx-go and linting in Go is considerably slower.

**Reply by meDavidNS** [reply]
Even with out frontend pipeline this takes multi-minute to lint around 170 projects in our repository, most of the time being taken by tailwind plugins for eslint.

**Comment by AndreasHerz** [comment]
Same issue here 👋🏼 
We would like to only publish directly affected projects to SonarCloud.
Dependencies don't matter…

**Reply by MattCCC** [reply]
+1. We ended up with a custom script and we're considering removal of nx because of that actually. If we still need a custom solution to check for dependencies then it's much less useful than we thought at the beginning. A pity.

**Reply by TheWallOfDucks** [reply]
Yeah we are in the exact same scenario using NX to publish results to SonarCloud...we end up having to publish results for lots of code that hasn't even changed.

**Comment by MattCCC** [comment]
What we did is basically a small node script that does git diff to take files and then we use nx affected with --files flag. Works like a charm. I still think it's something that nx affected should be doing out of the box.

**Reply by TheWallOfDucks** [reply]
Thanks for the idea @MattCCC . Sounds like that could work for my scenario too, but agree it's a shame it's not something available out of the box.

**Comment by chrismayverantos** [comment]
for precommit checks were doing something like this:
`npx nx affected:lint --files="$(git diff --cached --name-only -z | tr '\0' '\n' | paste -sd ',' -)"`

**Reply by meDavidNS** [reply]
That would still run the lint target for all the projects that have a dependency on the projects that are touched by the git diff, if I'm not mistaken? I think it would require manually filtering the nx graph structure to achieve only the directly affected projects.

**Reply by MattCCC** [reply]
> That would still run the lint target for all the projects that have a dependency on the projects that are touched by the git diff, if I'm not mistaken? I think it would require manually filtering the nx graph structure to achieve only the directly affected projects.

Yes, you are right. You would be much better off with a custom node script or an nx plugin instead. Git diff is the solution but you cannot combine it with affected. More likely with nx run-many.
